### ISSUE TITLE
[REVIEW]: mwarp1d: Manual one-dimensional data warping in Python and PyQt

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1870]
**Submitting author:** @0todd0000 (<a href="http://orcid.org/0000-0002-8292-7189">Todd Pataky</a>)
**Repository:** <a href="https://github.com/0todd0000/mwarp1d" target ="_blank">https://github.com/0todd0000/mwarp1d</a>
**Version:** 0.2
**Editor:** @trallard
**Reviewer:** @melund, @demotu
**Archive:** <a href="https://doi.org/10.5281/zenodo.3568446" target="_blank">10.5281/zenodo.3568446</a>

## Status

[![status](https://joss.theoj.org/papers/21abd77ffdb02ab669803273ba48c38b/status.svg)](https://joss.theoj.org/papers/21abd77ffdb02ab669803273ba48c38b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/21abd77ffdb02ab669803273ba48c38b"><img src="https://joss.theoj.org/papers/21abd77ffdb02ab669803273ba48c38b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/21abd77ffdb02ab669803273ba48c38b/status.svg)](https://joss.theoj.org/papers/21abd77ffdb02ab669803273ba48c38b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/0todd0000/mwarp1d" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@melund & @demotu, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @trallard know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/0todd0000/mwarp1d">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@0todd0000) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/0todd0000/mwarp1d">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@0todd0000) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @melund, @demotu it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01870/joss.01870/10.21105.joss.01870.pdf)

**Comment by trallard** [comment]
@melund @demotu, thanks for agreeing to review this submission, please feel free to carry your review using the checklist above.

If you have any questions please let me know here!

**Comment by 0todd0000** [comment]
I'm writing to request a status update... it has been 2 weeks since the review has started, but there has been activity from the reviewers in neither this thread, nor in the target repository, so I'm growing concerned that they may have not been properly added as reviewers. Please confirm that they they have indeed been added.

**Comment by kthyng** [comment]
Hi @0todd0000. I see that @melund checked boxes in the review above 7 days ago, so that reviewer was successfully added.

**Comment by melund** [comment]
Yes. Sorry about the slow progress @0todd0000 I did plan to play around with it last weekend but a lot of private stuff got in the way. I hope to find some time free time sunday or saturday evening.

**Comment by 0todd0000** [comment]
Dear Reviewers (@melund,@demotu), I'm very sorry to chase you again about this, but I'd like to request an expedited review. If at all possible, please submit your reviews this week. (It has now been over 6 weeks since submission, and for administrative reasons I really need review results before 1 December.) Thank you for considering this request.

**Comment by melund** [comment]
@0todd0000 I am OK with this. Did you see my comments this weekend. They are created as issues on [mwarp1d's issue tracker](https://github.com/0todd0000/mwarp1d/issues)

**Comment by 0todd0000** [comment]
@melund Oops!  I didn't see the issues, thank you for pointing them out. Will get on them now.

**Comment by melund** [comment]
No problem. I am in general positive about your tool. I have a little difficult seeing the use case for the the full manual warping. Maybe I am missing something. Is it possible to warp multiple segments along the curve or is it only one setting for the whole curve?  In what cases would you use it yourself? 

I do, however, think that a nice GUI tool for making landmark based warping could be really useful. Are there any ongoing projects in your group that use or plan to use (and cite) this tool? I think it would be good to mention them in the paper. 

So, in conclusion, if you address the issue I have about missing documentation source and lacking tests, then I think I can recommend this  publication.

**Comment by 0todd0000** [comment]
Thanks very much for the feedback!

(I've just responded to the three repository issues, including the Sphinx source code and automation. I had notifications turned off for some reason, so didn't see the issues before today, sorry about that!)

> I have a little difficult seeing the use case for the the full manual warping.

For landmark warping, the warp resolution (and indirectly: warp quality) is directly related to the number of landmarks. The manual mode permits high-resolution warping using mouse drags rather than mouse clicks.

> Is it possible to warp multiple segments along the curve or is it only one setting for the whole curve?

Yes, independent manual warps can be sequentially applied as indicated in the [manual warp screencast](http://www.spm1d.org/mwarp1d/quickstart.html#manual-warping-screencast) (at  around 0:40 ). These manual warps can affect an arbitrarily large (or small) regions of the 1D domain. In order to get finer control over local regions, the data must  be pre-interpolated to a finer resolution (e.g. 1000 points instead of 100).

> In what cases would you use it yourself?

We're not yet totally sure, but we're trying to answer this question objectively through experimentation. We're currently running a study to check performance differences between different methods (landmark, manual with low-resolution, manual with high-resolution), to try to identify the dataset characteristics that are best for each registration mode.

 > Are there any ongoing projects in your group that use or plan to use (and cite) this tool? I think it would be good to mention them in the paper.

Yes, the project above, but it's not yet close to a paper. We will certainly add it as a reference in the documentation if and when the study is published.

**Comment by demotu** [comment]
Hi Todd
The application is very interesting and potentially useful.
It's unfortunate that you don't have mwarp1d for Linux in your conda channel yet. So, you could mention that in your installation page and just tell the user what he/she already knows: install with python setup.py install.
The user manual and the example in overview.ipynb are helpful and clear.
The application is very complete and give lots of options for performing data warping.
The references don't appear in the paper.md but I can see them in the bib file, I guess it's ok.
Otherwise, as a biomechanicist working with these kind of problems, I didn't have any difficult in using your application.
Congratulations.

**Comment by melund** [comment]
>Yes, independent manual warps can be sequentially applied as indicated in the manual warp screencast (at around 0:40 ).

Thanks @0todd0000. I missed that when I played with the program.

**Comment by melund** [comment]
@0todd0000 has handled or answered all of my important comments: 

https://github.com/0todd0000/mwarp1d/issues/1
https://github.com/0todd0000/mwarp1d/issues/2
https://github.com/0todd0000/mwarp1d/issues/3

@trallard I can also recommend this for publication.

**Comment by 0todd0000** [comment]
@trallard I think both reviewers have recommended publication. Is there anything else I or the reviewers need to do for the review process?

**Comment by trallard** [comment]
Hey @0todd0000  I will start with my final editor duties. In the meantime, can you please complete the following actions?

- Final review of the paper for typos, author affiliations and missing DOIs
- Create a final release of the package and archive (i.e. Zenodo, Figshare)
- Indicate the final version of the package here
- Share the newly generated DOI here

**Comment by trallard** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by trallard** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1016/j.neuroimage.2004.06.018 is OK
- 10.1006/nimg.1997.0274 is OK
- 10.1214/15-STS524 is OK
- 10.1109/TEVC.2008.927706 is OK
- 10.1111/1467-9868.00129 is OK
- 10.1016/S0966-6362(00)00085-0 is OK
- 10.1111/biom.12963 is OK
- 10.1080/10618600.2019.1607744 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01870/joss.01870/10.21105.joss.01870.pdf)

**Comment by 0todd0000** [comment]
@melund 
@demotu 
Thank you very much for your time and comments!!!

**Comment by 0todd0000** [comment]
@trallard 

All final tasks complete. Here are the details:

* Final review of the paper for typos, author affiliations and missing DOIs

    * Two minor additions made to the paper for clarity. No typos found.
    * Author affiliations are correct.
    * No missing DOIs;  all DOIs re-checked.

* Create a final release of the package and archive (i.e. Zenodo, Figshare)

    * Done (Zenodo).

* Indicate the final version of the package here

     * Version:  0.2

* Share the newly generated DOI here:

    * [![DOI](https://zenodo.org/badge/190681079.svg)](https://zenodo.org/badge/latestdoi/190681079)

**Comment by 0todd0000** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01870/joss.01870/10.21105.joss.01870.pdf)

**Comment by 0todd0000** [comment]
@trallard 
The updated proof looks good.

**Comment by trallard** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1016/j.neuroimage.2004.06.018 is OK
- 10.1006/nimg.1997.0274 is OK
- 10.1214/15-STS524 is OK
- 10.1109/TEVC.2008.927706 is OK
- 10.1111/1467-9868.00129 is OK
- 10.1016/S0966-6362(00)00085-0 is OK
- 10.1111/biom.12963 is OK
- 10.1080/10618600.2019.1607744 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by trallard** [comment]
Thanks @0todd0000 just one note, the DOI in Zenodo points only to one single author (you in this case) but the paper has 2 more authors.

**Comment by 0todd0000** [comment]
@trallard 
Oops!  Sorry about that, the author list is now updated at Zenodo.

**Comment by trallard** [comment]
@whedon set version as 0.2

**Comment by trallard** [comment]
@whedon set 0.2 as version

**Comment by whedon** [comment]
OK. 0.2 is the version.

**Comment by trallard** [comment]
@whedon set 10.5281/zenodo.3568446 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3568446" target="_blank">10.5281/zenodo.3568446</a> is the archive.

**Comment by trallard** [comment]
I believe that this is ready to pass over to the last stage of the editorial process. 

@melund @demotu, thanks a lot for your time and contributions to JOSS through this review 🙌🏼

@openjournals/joss-eics I have concluded my editorial tasks and I would recommend this submission for publication. Thus I am sending it your way 👾🚀

**Comment by kyleniemeyer** [comment]
Hi @0todd0000, could you edit the title of the Zenodo entry to match the article?

**Comment by 0todd0000** [comment]
@kyleniemeyer Sure!  I've just updated the Zenodo entry's title.

**Comment by kyleniemeyer** [comment]
Hi @0todd0000, I'm just doing the final checks on the submission before publishing. Almost everything looks good, but I noticed one of your references was missing some info—could you add the volume, issue, and page number to the Ramsay and Li (1998) reference? I found: Volume 60, Issue 2, pages 351-363. Thanks!

**Comment by 0todd0000** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by 0todd0000** [comment]
Hi @kyleniemeyer, sorry about that!  I've updated the reference and have regenerated the PDF.

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01870/joss.01870/10.21105.joss.01870.pdf)

**Comment by kyleniemeyer** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1176

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1176, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1016/j.neuroimage.2004.06.018 is OK
- 10.1006/nimg.1997.0274 is OK
- 10.1214/15-STS524 is OK
- 10.1109/TEVC.2008.927706 is OK
- 10.1111/1467-9868.00129 is OK
- 10.1016/S0966-6362(00)00085-0 is OK
- 10.1111/biom.12963 is OK
- 10.1080/10618600.2019.1607744 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1206663943072178176) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1177
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01870](https://doi.org/10.21105/joss.01870)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congrats @0todd0000 on your article's publication in JOSS!

Thanks to @melund and @demotu for reviewing, and @trallard for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01870/status.svg)](https://doi.org/10.21105/joss.01870)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01870">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01870/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01870/status.svg
   :target: https://doi.org/10.21105/joss.01870
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01870/status.svg)](https://doi.org/10.21105/joss.01870)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
